### ISSUE TITLE
xOR Simple Example Neural Network

### ISSUE THREAD

**Issue by hpssjellis** [PAIR-code/deeplearnjs#846]
Can someone, preferably @tafsiri, write some pseudocode for an xOR solving neural network. Nothing fancy, just the basics:

Example:

[0,0]   =  0
[1,0]  =   1
[0,1]   = 1
[1,1]  =  0

It should be a small 3 layer network of: 2 inputs neurons, 3 hidden neurons and 1 output neuron. After the tensors are setup, it needs a train function (In the code below called propagate and a test function in the code below called activate). Fine results can be achieved with about 1000 training loops.

Give me a few weeks to research it, I could probably make a really inefficient version, which might work but would not be designed very well. So could someone give me a  starting point, something sensible in pseudocode that I can  then debug.

I have a working webpage written in [synaptic here](https://hpssjellis.github.io/deeplearnjs-javascript-examples/Unordered-javascript-examples/synaptic/neuralnetwork/xor01.html)

and the synaptic code for the page is following:

```
<script src="https://cdnjs.cloudflare.com/ajax/libs/synaptic/1.0.8/synaptic.min.js"></script>

<script>

// create the network
const { Layer, Network } = window.synaptic;

var inputLayer = new Layer(2);
var hiddenLayer = new Layer(3);
var outputLayer = new Layer(1);

inputLayer.project(hiddenLayer);
hiddenLayer.project(outputLayer);

var myNetwork = new Network({
input: inputLayer,
hidden: [hiddenLayer],
output: outputLayer
});

// train the network - learn XOR
var learningRate = .3;
for (var i = 0; i < 20000; i++)
{
// 0,0 => 0
myNetwork.activate([0,0]);
myNetwork.propagate(learningRate, [0]);

// 0,1 => 1
myNetwork.activate([0,1]);
myNetwork.propagate(learningRate, [1]);

// 1,0 => 1
myNetwork.activate([1,0]);
myNetwork.propagate(learningRate, [1]);

// 1,1 => 0
myNetwork.activate([1,1]);
myNetwork.propagate(learningRate, [0]);
}

</script>

<input type=button value=run onclick="{
                                      
  document.getElementById('myDiv01').innerHTML = 'xOR [0,0] = '+myNetwork.activate([0,0])+'<br>'
  document.getElementById('myDiv01').innerHTML += 'xOR [0,1] = '+myNetwork.activate([0,1])+'<br>'
  document.getElementById('myDiv01').innerHTML += 'xOR [1,0] = '+myNetwork.activate([1,0])+'<br>'
  document.getElementById('myDiv01').innerHTML += 'xOR [1,1] = '+myNetwork.activate([1,1])+'<br>'
}">
<div id="myDiv01">...</div>


```

And my github is here [https://github.com/hpssjellis/deeplearnjs-javascript-examples/tree/master/Unordered-javascript-examples/synaptic/neuralnetwork](https://github.com/hpssjellis/deeplearnjs-javascript-examples/tree/master/Unordered-javascript-examples/synaptic/neuralnetwork)

which has the synaptic non-minimum code

**Comment by ManrajGrover** [comment]
Hi Jeremy, did you have a look at #804? I believe it solves this issue.

**Comment by hpssjellis** [comment]
@ManrajGrover 

Wow that is awesome, and I did not see it. Actually was not even looking for it. 

This page is perfect 

[https://github.com/ManrajGrover/deeplearnjs/blob/xor-eager/demos/xor-eager/learn-xor.ts](https://github.com/ManrajGrover/deeplearnjs/blob/xor-eager/demos/xor-eager/learn-xor.ts)

**Comment by hpssjellis** [comment]
Thanks @ManrajGrover I got my javascript xOR webpage  working. The github is [here](https://github.com/hpssjellis/deeplearnjs-javascript-examples/tree/a5713442aec78f7aa3598018144b5339d700fa54/Unordered-javascript-examples/deeplearnjs-examples/xor-neural-network), code  [here](https://github.com/hpssjellis/deeplearnjs-javascript-examples/blob/a5713442aec78f7aa3598018144b5339d700fa54/Unordered-javascript-examples/deeplearnjs-examples/xor-neural-network/xor.html),  demo page [here](https://hpssjellis.github.io/deeplearnjs-javascript-examples/Unordered-javascript-examples/deeplearnjs-examples/xor-neural-network/xor.html)  

@tafsiri or anyone else: I have a few questions:

A: `const EPSILON = dl.scalar(1e-7); ` What is the point of EPSILON?

B: `const myOne = dl.scalar(1); ` I assume is the output neuron.

C: `const optimizer = dl.train.sgd(0.1);`  Anyone got any words of wisdom about optimizers for this neural network. Why 0.1 and why sgd?

D: `Math.sqrt(2.0 / prevLayerSize))`  Why? I realize they need to be initialized but why this equation?

E: 
```
function model(xs) {
    const hiddenLayer = dl.tidy( function ()  {
        return xs.matMul(fullyConnectedWeights1).add(fullyConnectedBias1).relu();
    });
    return hiddenLayer.matMul(fullyConnectedWeights2).add(fullyConnectedBias2).sigmoid();
}
```

relu and sigmoid do similar things, but if I reverse these everything messes up? Why?


..


..


![xor](https://user-images.githubusercontent.com/5605614/37324712-f0d14ff0-2647-11e8-8423-2f6ef45bb25f.png)


Reminder that  the code is [here](https://github.com/hpssjellis/deeplearnjs-javascript-examples/blob/a5713442aec78f7aa3598018144b5339d700fa54/Unordered-javascript-examples/deeplearnjs-examples/xor-neural-network/xor.html),

**Comment by ManrajGrover** [comment]
Hi @hpssjellis,

> A: const EPSILON = dl.scalar(1e-7); What is the point of EPSILON?
> 
> it occurs in the cost calculation
> 
> ```
> function calculateCost(y, output) {
>     return dl.add(y.mul(output.add(EPSILON).log()), myOne.sub(y).mul(myOne.sub(output).add(EPSILON).log())).sum().neg();
> }
> ```

It is to prevent `log` to tend to -infinity.

> 
> B: const myOne = dl.scalar(1); I assume is the output neuron.
> 

It is just a constant which is being used in multiple places. Hence, defined it once. It is not related to output neuron.

> C: const optimizer = dl.train.sgd(0.1); Anyone got any words of wisdom about optimizers for this neural network. Why 0.1 and why sgd?

`0.1` is the learning rate I used for this problem. Since it is a hyperparameter, you can tune it and see the effects. I found `0.1` to work for this problem. You can use different optimizers and see their effects. They may help learn and reach local minima much faster. I didn't experiment much with this.

> 
> D: Math.sqrt(2.0 / prevLayerSize)) Why? I realize they need to be initialized but why this equation?
> 

Reference [Weight Initialization: Calibrating the variances with 1/sqrt(n)](http://cs231n.github.io/neural-networks-2/#init)

> E:
> ```
> function model(xs) {
>     const hiddenLayer = dl.tidy( function ()  {
>         return xs.matMul(fullyConnectedWeights1).add(fullyConnectedBias1).relu();
>     });
>     return hiddenLayer.matMul(fullyConnectedWeights2).add(fullyConnectedBias2).sigmoid();
> }
> ```
> relu and sigmoid do similar things, but if I reverse these everything messes up? Why?

If you reverse them then the output will not be bounded between 0 and 1 because relu does not have an upper bound which is not suited for classification problems. Also, sigmoid should not be used in hidden layers because of vanishing gradient problem.

**Comment by hpssjellis** [comment]
Thanks @ManrajGrover did not know you were the original maker of the code. I really appreciate the help. Learnt more from your replies than tons of googling about Machine Learning.


D: `Math.sqrt(2.0 / prevLayerSize))` Why? I realize they need to be initialized but why this equation?

From your[ link](http://cs231n.github.io/neural-networks-2/#init) I found this answer for D:

> Calibrating the variances with 1/sqrt(n). One problem with the above suggestion is that the distribution of the outputs from a randomly initialized neuron has a variance that grows with the number of inputs. It turns out that we can normalize the variance of each neuron’s output to 1 by scaling its weight vector by the square root of its fan-in (i.e. its number of inputs). That is, the recommended heuristic is to initialize each neuron’s weight vector as: w = np.random.randn(n) / sqrt(n), where n is the number of its inputs. This ensures that all neurons in the network initially have approximately the same output distribution and empirically improves the rate of convergence.

**Comment by ManrajGrover** [comment]
@hpssjellis You should follow deeplearnjs twitter channel where you can find various demos being shared.

**Comment by codeKonami** [comment]
I'm trying to write a xOR solving neural network with tfjs. I've got something working-ish but it's taking a lot of time to train. It needs about 10000 epochs to reach a 0.01 loss.
Could someone have a quick look to help me out?

Thanks

https://jsbin.com/waqipeh/edit?html,console

I'm copy pasting the code below too:

```js
const model = tf.sequential();
model.add(tf.layers.dense({
     units: 10,
     inputShape: [2],
     activation: 'relu',
   }));
model.add(tf.layers.dense({
     units: 1,
     activation: 'sigmoid'
   }));

model.compile({optimizer: 'sgd', loss: 'meanSquaredError'});
 
const Xs = tf.tensor2d([
  [0, 0],
  [0, 1],
  [1, 0],
  [1, 1]
], [4, 2]);

const Ys = tf.tensor2d([
  [0],
  [1],
  [1],
  [0]
], [4, 1]);

async function train() {
  const h = await model.fit(Xs, Ys, {
     epochs: 1000,
     batchSize: 4,
     callbacks: {
       onEpochEnd: async (epoch, logs) => {
         if (epoch % 100 === 0) {
           console.log(`epoch ${epoch} -- loss: ${logs.loss}`);
         }
       }
     }
  });
}

(async () => {
  console.log('Training...')
  await train();
  console.log(`[0, 0] should return 0 => ${await model.predict(tf.tensor2d([0, 0], [1, 2])).data()}`);
  console.log(`[0, 1] should return 1 => ${await model.predict(tf.tensor2d([0, 1], [1, 2])).data()}`);
  console.log(`[1, 0] should return 1 => ${await model.predict(tf.tensor2d([1, 0], [1, 2])).data()}`);
  console.log(`[1, 1] should return 0 => ${await model.predict(tf.tensor2d([1, 1], [1, 2])).data()}`);
})()
```
